File: app/src/main/java/com/polok/eubmanagement/presentation/notice/noticeadd/NoticeAddViewModel.java
package com.polok.eubmanagement.presentation.notice.noticeadd;

import static com.polok.eubmanagement.util.Extension.showErrorOnUi;
import android.widget.EditText;
import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DatabaseReference;
import com.polok.eubmanagement.base.BaseViewModel;
import com.polok.eubmanagement.base.model.OnNavigate;
import com.polok.eubmanagement.firebase.FirebaseDataRef;
import com.polok.eubmanagement.model.NoticeData;
import com.polok.eubmanagement.util.Extension;

public class NoticeAddViewModel extends BaseViewModel {

    public void validateNoticeInputAndUploadToFirebase(
            EditText noticeTitleEt, EditText noticeDetailsEt
    ) {
        if (noticeTitleEt.getText().toString().isEmpty()) {
            showErrorOnUi(noticeTitleEt, "Enter Notice Title");
            return;
        }
        if (noticeDetailsEt.getText().toString().isEmpty()) {
            showErrorOnUi(noticeDetailsEt, "Enter Notice Details");
            return;
        }
        uploadNoticeTOFirebase(
                noticeTitleEt.getText().toString(),
                noticeDetailsEt.getText().toString(),
                Extension.getCurrentDate()
        );
    }

    private void uploadNoticeTOFirebase(String title, String details, String date) {
        fireLoadingEvent(true);
        DatabaseReference pushNoticeRef = FirebaseDataRef.provideNoticeRef().push();
        pushNoticeRef.setValue(
                new NoticeData(title,details,date,pushNoticeRef.getKey().toString())
        ).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isComplete()) {
                    fireMessageEvent("Notice Added Successfully");
                    fireNavigateEvent(new OnNavigate(1));
                } else fireMessageEvent(task.getException().getLocalizedMessage());
                fireLoadingEvent(false);
            }
        });
    }
}

File: app/src/main/java/com/polok/eubmanagement/presentation/notice/noticeadd/NoticeAddViewModel.kt
package com.polok.eubmanagement.presentation.notice.noticeadd

import android.widget.EditText
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.google.firebase.database.DatabaseReference
import com.polok.eubmanagement.base.BaseViewModel
import com.polok.eubmanagement.firebase.FirebaseDataRef.provideNoticeRef
import com.polok.eubmanagement.model.NoticeData
import com.polok.eubmanagement.util.getCurrentDate
import com.polok.eubmanagement.util.showErrorOnUi

class NoticeAddViewModel : BaseViewModel() {

    fun validateNoticeInputAndUploadToFirebase(
        noticeTitleEt: EditText, noticeDetailsEt: EditText
    ) {
        if (noticeTitleEt.getText().toString().isEmpty()) {
            noticeTitleEt.showErrorOnUi("Enter Notice Title")
            return
        }
        if (noticeDetailsEt.getText().toString().isEmpty()) {
            noticeDetailsEt.showErrorOnUi("Enter Notice Details")
            return
        }
        uploadNoticeTOFirebase(
            noticeTitleEt.getText().toString(),
            noticeDetailsEt.getText().toString(),
            getCurrentDate()
        )
    }

    private fun uploadNoticeTOFirebase(title: String, details: String, date: String) {
        fireLoadingEvent(true)
        val dbPushRef: DatabaseReference? = provideNoticeRef()?.push()
        dbPushRef?.setValue(
            NoticeData(
                title = title,
                details = details,
                createdAt = date,
                key = dbPushRef.key
            )
        )?.addOnCompleteListener { task ->
            if (task.isComplete) {
                fireMessageEvent("Notice Submitted Successfully")
                fireNavigateEvent(0, null)
            } else {
                fireMessageEvent(task.exception?.localizedMessage)
            }
            fireLoadingEvent(false)
        }
    }

    class Factory : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return NoticeAddViewModel() as T
        }
    }
}